src/components/PageContent/ContentObserver/ContentObserver.ts

```diff
@@ -0,0 +1,49 @@
+import { useEffect } from "preact/hooks"
+import { useCurrentIds } from "~/hooks/currentIds/useCurrentIds"
+import { currentIds } from "~/hooks/currentIds/idStore"
+import { MarkdownHeading } from "astro"
+import { useStickyHeader } from "~/hooks/stickyHeader/useStickyHeader"
+
+export const ContentObserver = ({ headings, shouldUpdate }: { headings: MarkdownHeading[]; shouldUpdate: number }) => {
+  const { setCurrentIds } = useCurrentIds()
+  const { setStickyHeader } = useStickyHeader()
+
+  useEffect(() => {
+    const observerCallback: IntersectionObserverCallback = (entries) => {
+      let stickyHeaderSet = false
+      const ids: Record<string, boolean> = currentIds.get()
+      for (const entry of entries) {
+        if (!entry.isIntersecting && entry.target.id === "overview") {
+          setStickyHeader("")
+        }
+        ids[entry.target.id] = entry.isIntersecting
+        if (
+          !stickyHeaderSet &&
+          entry.isIntersecting &&
+          entry.target.firstElementChild?.textContent &&
+          ["H1", "H2"].includes(entry.target.firstElementChild.nodeName)
+        ) {
+          stickyHeaderSet = true
+          setStickyHeader(
+            entry.target.firstElementChild.id === "overview" ? "Overview" : entry.target.firstElementChild.textContent
+          )
+        }
```

Comment:
While I agree with this idea, doing this leads to some typescript confusion where it requires me casting `firstElementChild.textContent` to string, since it otherwise interprets it as `null | string`. This might just be an issue with Typescript. But it ends up being more readable the way it is